ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -311,20 +445,41 @@ def add_file(self, fileGrp, mimetype=None, url=None, ID=None, pageId=None, force
         if not REGEX_FILE_ID.fullmatch(fileGrp):
             raise ValueError("Invalid syntax for mets:fileGrp/@USE %s (not an xs:ID)" % fileGrp)
         log = getLogger('ocrd_models.ocrd_mets.add_file')
+
+        """
+        # Note: we do not benefit enough from having 
+        # a separate cache for fileGrp elements
+
+        if self._cache_flag: 
+            if fileGrp in self._fileGrp_cache:
+                el_fileGrp = self._fileGrp_cache[fileGrp]
+        """
+
         el_fileGrp = self.add_file_group(fileGrp)
         if not ignore:
-            mets_file = next(self.find_files(ID=ID), None)
+            # Since we are sure that fileGrp parameter is set,
+            # we could send that parameter to find_files for direct search
+            mets_file = next(self.find_files(ID=ID, fileGrp=fileGrp), None)
             if mets_file:
                 if mets_file.fileGrp == fileGrp and \
                    mets_file.pageId == pageId and \
                    mets_file.mimetype == mimetype:
                     if not force:
                         raise FileExistsError(f"A file with ID=={ID} already exists {mets_file} and neither force nor ignore are set")
-                    self.remove_file(ID=ID)
+                    self.remove_file(ID=ID, fileGrp=fileGrp)
```

Comment:
`ID` must be unique, I see no benefit in adding `fileGrp` to the call.

Reply:
`remove_file()` calls `find_files()` inside. So, by passing the fileGrp as well we reduce the candidates size inside `find_files()`. The candidates list hold only the IDs inside a specific fileGrp and not all IDs. Without the suggested change we iterate over all IDs, with the suggested change we iterate only over some sub-list of the IDs.